doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
+   and some of the types :n:`A__1, ..., A__n` may become new
+   subgoals.  It won't find occurrences inside `forall`; use :tacn:`setoid_rewrite`
+   if you want to find such occurrences.
 
-   .. tacv:: rewrite @term by @tactic
+   :n:`{+, @oriented_rewriter }`
+     The :n:`@oriented_rewriter`\s are applied sequentially
+     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
+     the tactic fails.
 
-      Use tactic to completely solve the side-conditions arising from the
-      :tacn:`rewrite`.
+   :n:`{? {| -> | <- } }`
+     For `->` (the default), :n:`term__1` is rewritten
+     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term__1`.
 
-   .. tacv:: rewrite {+, @orientation @term} {? in @ident }
+   :n:`{? @natural } {? {| ? | ! } }`
+     :n:`natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
+     is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number
+     of rewrites to perform.
 
-      Is equivalent to the `n` successive tactics :n:`{+; rewrite @term}`, each one
-      working on the first subgoal generated by the previous one. An :production:`orientation`
-      ``->`` or ``<-`` can be inserted before each :token:`term` to rewrite.  One
-      unique clause can be added at the end after the keyword in; it will then
-      affect all rewrite operations.
+     '?' performs the rewrite as many times as possible (possibly zero times).
+     This form never fails.  '!' performs the rewrite as many times as possible
+     and at least once.  The tactic fails if the requested number of rewrites can't
+     be performed.
 
-   In all forms of rewrite described above, a :token:`term` to rewrite can be
-   immediately prefixed by one of the following modifiers:
+   :n:`@occurrences`
+     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
+     any of them can be rewritten.  If not specified, only the first occurrence
+     in the conclusion is replaced.
 
-   + `?` : the tactic :n:`rewrite ?@term` performs the rewrite of :token:`term` as many
-     times as possible (perhaps zero time). This form never fails.
-   + :n:`@natural?` : works similarly, except that it will do at most :token:`natural` rewrites.
-   + `!` : works as `?`, except that at least one rewrite should succeed, otherwise
-     the tactic fails.
-   + :n:`@natural!` (or simply :n:`@natural`) : precisely :token:`natural` rewrites of :token:`term` will be done,
-     leading to failure if these :token:`natural` rewrites are not possible.
+     If :n:`at @occs_nums` is specified, rewriting is always done with setoid
+     rewriting, even for Leibniz’s equality, so you must
+     ``Require Setoid`` to use this form.
 
-   .. tacv:: erewrite @term
-      :name: erewrite
+     .. todo PR occurrences are not found within forall.  Is this specific to
+        rewrite or a general limitation of occurrence finding?
```

Comment:
It is probably not entirely specific to this tactic but not general either. It depends on the tactic.

Reply:
OK.  Is this a necessary condition or would it be worthwhile to relax the condition?

Reply:
In rewrite, it's not strictly necessary since `setoid_rewrite` proves that it's possible to do differently (though it may be more complicated and may lead to error messages that are harder to understand). But I think that there certainly are tactics for which it just doesn't make sense to select an occurrence under binders.